server/bin/pbench-cull-unpacked-tarballs.py

```diff
@@ -0,0 +1,457 @@
+    which point to the given incoming tar ball directory.
+
+    Any errors encountered will be logged.
+
+    Returns 0 on success, > 0 if any errors encountered
+    """
+    errors = 0
+    actions_taken = []
+    for dir_n, subdir_l, file_l in os.walk(tgt_p):
+        # NOTE: we ignore any files found, as pbench-audit-server should flag
+        # subject objects.
+        for subdir_n in subdir_l:
+            full_p = os.path.join(dir_n, subdir_n)
+            try:
+                link = os.readlink(full_p)
+            except OSError:
+                continue
+            if os.path.realpath(link) == tb_incoming_dir:
+                # FIXME: Remember the path to the removed symbolic link, and
+                # remove any parent directories that become empty as a result
+                # of the symbolic link removal.
+                if dry_run:
+                    print(f"rm {full_p}")
+                else:
+                    try:
+                        os.unlink(full_p)
+                    except OSError as exc:
+                        logger.error("Failed to remove symlink '{}': {}", full_p, exc)
+                        errors += 1
+                    else:
+                        logger.debug("Removed symlink '{}'", full_p)
+                        actions_taken.append(("rm", full_p))
+    # FIXME: By removing any symlinks, the directory containing that symlink
+    # may now be empty.  And it could be the entire prefix chain could be
+    # removed.  However, we don't want to remove prefixes, users, or
+    # controllers at this point since that might create a race condition for
+    # unpack tar balls (unpacking of a tar ball with a result prefix hierarchy
+    # which matches another tar being removed from the same controller).
+    return errors, actions_taken
+
+
+def remove_unpacked(tb_incoming_dir, controller_name, results, users, logger, dry_run):
+    """remove_unpacked - Remove the unpacked tar ball directory from the
+    INCOMING tree and all symbolic links to that directory from the RESULTS
+    and USERS trees.
+
+    The symbolic links in the RESULTS and USERS (if any) trees are removed
+    first, then the directory in the INCOMING tree is removed.
+
+    The `tb_incoming_dir` should be a fully resolved path including the tar
+    ball directory name, allowing us to resolve the symbolic link being
+    considered to compare for equality.
+
+    Any errors encountered while removing items will result in those errors
+    being logged.  All removals stop on the first error encountered, so the
+    removal process could leave the unpacked tar ball in a partially deleted
+    state.
+
+    Returns 0 on success, > 0 if any errors encountered.
+    """
+    duration = 0
+    if not dry_run:
+        logger.info("Began removing unpacked tar ball directory, '{}'", tb_incoming_dir)
+        start = pbench._time()
+
+    # Walk the results hierarchy for the tar ball's controller to find all
```

Comment:
Can there be more than one? Isn't that one the specified prefix? Walking the hierarchy is probably warranted to make sure that nothing is left behind, but it should be possible to retrieve the prefix (either from the prefix file in the archive or from the metadata.log) and target that only - right?

Reply:
It certainly seems like there can be only one link.  But given this age of all the existing unpacked tar ball and successive versions of server software, this seemed like the safest way to ensure all things are cleaned up.

Reply:
Agreed.